[Edit] - Pandas drop_duplicates (#6687)
Codecademy/docs: content/pandas/concepts/dataframe/terms/drop-duplicates/drop-duplicates.md

---
Title: '.drop_duplicates()'
Description: 'Removes duplicate rows from a DataFrame based on specified columns.'
Subjects:
  - 'Computer Science'
  - 'Data Science'
Tags:
  - 'Data'
  - 'Data Structures'
  - 'Data Types'
  - 'Pandas'
  - 'Python'
CatalogContent:
  - 'learn-python-3'
  - 'paths/data-science'
---

> The **`.drop_duplicates()`** method in Pandas is used to remove duplicate rows from a DataFrame. This method identifies and eliminates duplicated rows across all columns or based on specified columns, helping to clean and preprocess data for analysis. When working with real-world datasets, duplicate records often arise due to data entry errors, merging datasets, or data collection issues, making `.drop_duplicates()` an essential tool in a data scientist's toolkit.
> 
> By default, `.drop_duplicates()` retains the first occurrence of each duplicate row and removes all subsequent duplicates. This method is particularly useful in data cleaning pipelines, ensuring data integrity for accurate analysis, and reducing memory usage by eliminating redundant information.

Reviewer: ```suggestion
The **`.drop_duplicates()`** method in Pandas is used to remove duplicate rows from a [`DataFrame`](https://www.codecademy.com/resources/docs/pandas/dataframe). This method identifies and eliminates duplicate rows across all columns or based on specified columns, helping to clean and preprocess data for analysis. When working with real-world datasets, duplicate records often arise due to data entry errors, merging datasets, or data collection issues, making `.drop_duplicates()` an essential tool in a data scientist's toolkit.

By default, `.drop_duplicates()` retains the first occurrence of each duplicate row and removes all subsequent duplicates. This method is particularly useful in data cleaning pipelines, ensuring data integrity for accurate analysis, and reducing memory usage by eliminating redundant information.
```

## Syntax

```pseudo
DataFrame.drop_duplicates(subset=None, keep='first', inplace=False, ignore_index=False)
```

> **Parameters:**
> 
> - `subset`: A column label or a list of column labels to consider for identifying duplicates. Default is `None`, which uses all columns.
> - `keep`: Determines which duplicates to keep. Options include:
>   - `'first'`: Keep the first occurrence (default)
>   - `'last'`: Keep the last occurrence
>   - `False`: Drop all duplicates
> - `inplace`: Boolean value that determines whether to modify the `DataFrame` directly (`True`) or return a new `DataFrame` with duplicates removed (`False`). Default is `False`.
> - `ignore_index`: Boolean value that, when `True`, resets the index of the resulting DataFrame. Default is `False`.
> 
> **Return value:**
> 
> Returns a DataFrame with duplicate rows removed or `None` if `inplace=True`.

Reviewer: ```suggestion
**Parameters:**

- `subset`: A column label or a list of column labels to consider for identifying duplicates. Default is `None`, which uses all columns.
- `keep`: Determines which duplicates to keep. Options include:
  - `'first'`: Keep the first occurrence (default)
  - `'last'`: Keep the last occurrence
  - `False`: Drop all duplicates
- `inplace`: Boolean value that determines whether to modify the `DataFrame` directly (`True`) or return a new `DataFrame` with duplicates removed (`False`). Default is `False`.
- `ignore_index`: Boolean value that, when `True`, resets the index of the resulting `DataFrame`. Default is `False`.

**Return value:**

Returns a `DataFrame` with duplicate rows removed or `None` if `inplace=True`.
```

## Example 1: Basic Duplicate Removal

The following example demonstrates how to remove duplicate rows from a DataFrame containing customer information:

```py
# Import pandas library
import pandas as pd

# Create a sample DataFrame with duplicated rows
data = {
  'Name': ['Alice', 'Bob', 'Alice', 'David'],
  'Age': [25, 30, 25, 40],
  'City': ['NY', 'LA', 'NY', 'Chicago']
}

# Create DataFrame
df = pd.DataFrame(data)

# Display the original DataFrame
print("Original DataFrame:")
print(df)

# Remove duplicates and create a new DataFrame
unique_df = df.drop_duplicates()

# Display the DataFrame with duplicates removed
print("\nDataFrame after removing duplicates:")
print(unique_df)
```

> The output generated by the code will be:

Reviewer: ```suggestion
The output generated by the above code will be:
```

print("\nKeeping first readings of each sensor (keep='first'):")
print(first_readings)

# Keep last occurrence of each sensor (latest reading)
last_readings = sensor_data.drop_duplicates(subset=['SensorID'], keep='last')
print("\nKeeping last readings of each sensor (keep='last'):")
print(last_readings)

# Drop all duplicates (keep only sensors with one reading)
unique_sensors = sensor_data.drop_duplicates(subset=['SensorID'], keep=False)
print("\nKeeping only sensors with unique readings (keep=False):")
print(unique_sensors)

# Modify the original DataFrame in place
sensor_data_copy = sensor_data.copy()
sensor_data_copy.drop_duplicates(subset=['SensorID'], inplace=True)
print("\nModifying DataFrame in place (inplace=True):")
print(sensor_data_copy)
```

This example illustrates the different `keep` options:

- With `keep='first'`, the earliest reading from each sensor is retained
- With `keep='last'`, the latest reading from each sensor is kept
- With `keep=False`, no duplicates are retained (empty DataFrame since all sensors have multiple readings)
- Using `inplace=True` modifies the original DataFrame directly

> ## Frequently Asked Questions
> 
> ### 1. Does `.drop_duplicates()` consider all columns by default?
> 
> Yes, by default, `.drop_duplicates()` considers all columns when identifying duplicates. To focus on specific columns, use the `subset` parameter.
> 
> ### 2. How does `.drop_duplicates()` handle `NaN` values?
> 
> `NaN` values are considered equal to other `NaN` values when comparing rows for duplicates.
> 
> ### 3. What is the time complexity of `.drop_duplicates()`?
> 
> The time complexity is approximately O(n), where n is the number of rows, although it can vary based on the number of columns being considered.
> 
> ### 4. Can I use `.drop_duplicates()` with multiIndex `DataFrames`?
> 
> Yes, `.drop_duplicates()` works with multiIndex `DataFrames`. You can specify index levels in the `subset` parameter if needed.
> 
> ### 5. How does the `ignore_index` parameter work?
> 
> When `ignore_index=True`, the resulting `DataFrame` will have a new index with consecutive integers starting from 0, regardless of the original index values.

Reviewer: ```suggestion
## Frequently Asked Questions

### 1. Does `.drop_duplicates()` consider all columns by default?

Yes, by default, `.drop_duplicates()` considers all columns when identifying duplicates. To focus on specific columns, use the `subset` parameter.

### 2. How does `.drop_duplicates()` handle `NaN` values?

`NaN` values are considered equal to other `NaN` values when comparing rows for duplicates.

### 3. What is the time complexity of `.drop_duplicates()`?

The time complexity is approximately `O(n)`, where n is the number of rows, although it can vary based on the number of columns being considered.

### 4. Can I use `.drop_duplicates()` with multiIndex `DataFrames`?

Yes, `.drop_duplicates()` works with multiIndex `DataFrames`. You can specify index levels in the `subset` parameter if needed.

### 5. How does the `ignore_index` parameter work?

When `ignore_index=True`, the resulting `DataFrame` will have a new index with consecutive integers starting from 0, regardless of the original index values.
```
